lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,318 @@
+// Copyright (C) 2016 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package fswatcher
+
+import (
+	"errors"
+	"fmt"
+	"github.com/zillode/notify"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/syncthing/syncthing/lib/events"
+	"github.com/syncthing/syncthing/lib/ignore"
+)
+
+type FsEvent struct {
+	path string
+	time time.Time
+}
+
+type FsEventsBatch map[string]*FsEvent
+
+type FsWatcher struct {
+	folderPath      string
+	notifyModelChan chan<- FsEventsBatch
+	// All detected and to be scanned events.
+	fsEvents              FsEventsBatch
+	fsEventChan           chan notify.EventInfo
+	WatchingFs            bool
+	notifyDelay           time.Duration
+	slowNotifyDelay       time.Duration
+	notifyTimer           *time.Timer
+	notifyTimerNeedsReset bool
+	inProgress            map[string]struct{}
+	folderID              string
+	ignores               *ignore.Matcher
+	ignoresUpdate         chan *ignore.Matcher
+	// Keeps track the events that are tracked within a directory for event
+	// aggregation. The directory itself is not (yet) to be scanned.
+	trackedDirs map[string]FsEventsBatch
+	stop        chan struct{}
+}
+
+const (
+	fastNotifyDelay = time.Duration(500) * time.Millisecond
+	maxFiles        = 512
+	maxFilesPerDir  = 128
+)
+
+func NewFsWatcher(folderPath string, folderID string, ignores *ignore.Matcher,
+	slowNotifyDelayS int) *FsWatcher {
+	if slowNotifyDelayS == 0 {
+		slowNotifyDelayS = 60 * 60 * 24
+	}
+	return &FsWatcher{
+		folderPath:            folderPath,
+		notifyModelChan:       nil,
+		fsEvents:              make(FsEventsBatch),
+		fsEventChan:           nil,
+		WatchingFs:            false,
+		notifyDelay:           fastNotifyDelay,
+		slowNotifyDelay:       time.Duration(slowNotifyDelayS) * time.Second,
+		notifyTimerNeedsReset: false,
+		inProgress:            make(map[string]struct{}),
+		folderID:              folderID,
+		ignores:               ignores,
+		ignoresUpdate:         make(chan *ignore.Matcher),
+		trackedDirs:           make(map[string]FsEventsBatch),
+		stop:                  make(chan struct{}),
+	}
+}
+
+func (watcher *FsWatcher) StartWatchingFilesystem() (<-chan FsEventsBatch, error) {
+	fsEventChan, err := watcher.setupNotifications()
+	notifyModelChan := make(chan FsEventsBatch)
+	watcher.notifyModelChan = notifyModelChan
+	if err == nil {
```

Comment:
Unless there is very good reason otherwise, please always prefer to do the `err != nil` check with early return and letting the success path be the default one.